File: source/downloads/code/LeetCode/Java/Maximum-Subarray.java
public class Solution {
    public int maxSubArray(int[] nums) {
        int n = nums.length;
        if (n==0)
            return 0;
        int maxSoFar = nums[0];
        int maxEnding = nums[0];
        for (int i=1;i<n;i++){
            maxEnding = Math.max(nums[i],nums[i]+maxEnding);
            maxSoFar = Math.max(maxSoFar, maxEnding); 
        }        
        return maxSoFar;
    }
}

File: downloads/code/LeetCode/Python/Maximum-Subarray.py
class Solution(object):
    def maxSubArray(self, nums):
        maxSum = nums[0]
        sum = 0
        for i in nums:
            # if sum > 0, then add i, no matter i>0 or i<0
            if sum > 0:
                sum += i
            else:
            # if sum < 0, then make sum start over from i
                sum = i
            if sum > maxSum:
                maxSum = sum
        return maxSum
        
        
